crates/ruff_python_semantic/src/nodes.rs

```diff
@@ -0,0 +1,136 @@
+use std::ops::Index;
+
+use ruff_index::{newtype_index, IndexVec};
+use ruff_python_ast::{Expr, Ranged, Stmt};
+use ruff_text_size::TextRange;
+
+use crate::BranchId;
+
+/// Id uniquely identifying an AST node in a program.
+///
+/// Using a `u32` is sufficient because Ruff only supports parsing documents with a size of max
+/// `u32::max` and it is impossible to have more nodes than characters in the file. We use a
+/// `NonZeroU32` to take advantage of memory layout optimizations.
+#[newtype_index]
+#[derive(Ord, PartialOrd)]
+pub struct NodeId;
+
+/// An AST node in a program, along with a pointer to its parent node (if any).
+#[derive(Debug)]
+struct NodeWithParent<'a> {
+    /// A pointer to the AST node.
+    node: NodeRef<'a>,
+    /// The ID of the parent of this node, if any.
+    parent: Option<NodeId>,
+    /// The branch ID of this node, if any.
+    branch: Option<BranchId>,
+}
+
+/// The nodes of a program indexed by [`NodeId`]
+#[derive(Debug, Default)]
+pub struct Nodes<'a> {
+    nodes: IndexVec<NodeId, NodeWithParent<'a>>,
+}
+
+impl<'a> Nodes<'a> {
+    /// Inserts a new AST node into the tree and returns its unique ID.
+    pub(crate) fn insert(
+        &mut self,
+        node: NodeRef<'a>,
+        parent: Option<NodeId>,
+        branch: Option<BranchId>,
+    ) -> NodeId {
+        self.nodes.push(NodeWithParent {
+            node,
+            parent,
+            branch,
+        })
+    }
+
+    /// Return the [`NodeId`] of the parent node.
+    #[inline]
+    pub fn parent_id(&self, node_id: NodeId) -> Option<NodeId> {
+        self.nodes[node_id].parent
+    }
+
+    /// Return the [`BranchId`] of the branch node.
+    #[inline]
+    pub(crate) fn branch_id(&self, node_id: NodeId) -> Option<BranchId> {
+        self.nodes[node_id].branch
+    }
+
+    /// Returns an iterator over all [`NodeId`] ancestors, starting from the given [`NodeId`].
+    pub(crate) fn ancestor_ids(&self, node_id: NodeId) -> impl Iterator<Item = NodeId> + '_ {
+        std::iter::successors(Some(node_id), |&node_id| self.nodes[node_id].parent)
+    }
+}
+
+impl<'a> Index<NodeId> for Nodes<'a> {
+    type Output = NodeRef<'a>;
+
+    #[inline]
+    fn index(&self, index: NodeId) -> &Self::Output {
+        &self.nodes[index].node
+    }
+}
+
+/// A reference to an AST node. Like [`ruff_python_ast::node::AnyNodeRef`], but wraps the node
+/// itself (like [`Stmt`]) rather than the narrowed type (like [`ruff_python_ast::StmtAssign`]).
+///
+/// TODO(charlie): Replace with [`ruff_python_ast::node::AnyNodeRef`]. This requires migrating
+/// the rest of the codebase to use [`ruff_python_ast::node::AnyNodeRef`] and related abstractions,
+/// like [`ruff_python_ast::expression::ExpressionRef`] instead of [`Expr`].
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum NodeRef<'a> {
+    Stmt(&'a Stmt),
+    Expr(&'a Expr),
+}
```

Comment:
@MichaReiser - I really want this API to be capable of returning `&Stmt` and `&Expr` so that we can separate this refactor (and some follow-up refactors) from sweeping changes to the Ruff internals to use `ExpressionRef` etc. everywhere, which we should probably do but don't need to be coupled to this change. I can't figure out how to do this with `AnyNodeRef` -- is it possible?

Reply:
This, unfortunately, isn't possible (and the reason why `ExpressionRef` exists). It is impossible to go from e.g. a `&StmtIf` to `&Stmt::If`

Reply:
Can you expand on why / how this motivates `ExpressionRef`? My understanding is that `ExpressionRef` has the same problem -- you can't go from `ExpressionRef::BoolOp` to `&Expr`, only `&ExprBoolOp`.

Reply:
It's correct that you can't go from `ExpressionRef` to `&Expr`. But `ExpressionRef` allows you to go from `AnyNodeRef` to `ExprRef`. Meaning, `ExpressionRef` gives you a union that is limited to the variants of `Expr`, but can be converted to from any node.

Reply:
Okay this makes sense, but it still has the same problem, which is that if I use `AnyNodeRef` here, I have to make extensive changes to the rest of the codebase to use `StatementRef`, `ExpressionRef`, etc. instead of `&Stmt`, `&Expr`, etc.